card/src/main/java/com/adyen/checkout/card/DefaultCardDelegate.kt

```diff
@@ -0,0 +1,597 @@
+import com.adyen.checkout.card.util.CardValidationUtils
+import com.adyen.checkout.card.util.DetectedCardTypesUtils
+import com.adyen.checkout.card.util.InstallmentUtils
+import com.adyen.checkout.card.util.KcpValidationUtils
+import com.adyen.checkout.card.util.SocialSecurityNumberUtils
+import com.adyen.checkout.components.base.AddressVisibility
+import com.adyen.checkout.components.model.paymentmethods.PaymentMethod
+import com.adyen.checkout.components.model.payments.request.CardPaymentMethod
+import com.adyen.checkout.components.model.payments.request.PaymentComponentData
+import com.adyen.checkout.components.repository.PublicKeyRepository
+import com.adyen.checkout.components.ui.FieldState
+import com.adyen.checkout.components.ui.Validation
+import com.adyen.checkout.components.util.PaymentMethodTypes
+import com.adyen.checkout.core.exception.CheckoutException
+import com.adyen.checkout.core.exception.ComponentException
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import com.adyen.checkout.cse.CardEncrypter
+import com.adyen.checkout.cse.EncryptedCard
+import com.adyen.checkout.cse.GenericEncrypter
+import com.adyen.checkout.cse.UnencryptedCard
+import com.adyen.checkout.cse.exception.EncryptionException
+import com.adyen.threeds2.ThreeDS2Service
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+
+@Suppress("LongParameterList", "TooManyFunctions")
+class DefaultCardDelegate(
+    private val publicKeyRepository: PublicKeyRepository,
+    private val configuration: CardConfiguration,
+    private val paymentMethod: PaymentMethod,
+    private val addressRepository: AddressRepository,
+    private val detectCardTypeRepository: DetectCardTypeRepository,
+    private val cardValidationMapper: CardValidationMapper,
+    private val cardEncrypter: CardEncrypter,
+    private val genericEncrypter: GenericEncrypter,
+) : CardDelegate {
+
+    override val inputData: CardInputData = CardInputData()
+
+    private var publicKey: String? = null
+
+    private val _outputDataFlow = MutableStateFlow<CardOutputData?>(null)
+    override val outputDataFlow: Flow<CardOutputData?> = _outputDataFlow
+
+    private val outputData
+        get() = _outputDataFlow.value
+
+    private val _componentStateFlow = MutableStateFlow<CardComponentState?>(null)
+    override val componentStateFlow: Flow<CardComponentState?> = _componentStateFlow
+
+    private val _exceptionFlow = MutableSharedFlow<CheckoutException>(0, 1, BufferOverflow.DROP_OLDEST)
+    override val exceptionFlow: Flow<CheckoutException> = _exceptionFlow
+
+    private var coroutineScope: CoroutineScope? = null
```

Comment:
I don't like all the null checks for this field in the methods, so maybe we can do something like this:
```
private var _coroutineScope: CoroutineScope? = null
private val coroutineScope: CoroutineScope get() = requireNotNull(_coroutineScope)
```
This should never go wrong and if it does something is implemented incorrectly